kore/src/Kore/Exec.hs

```diff
@@ -638,6 +654,51 @@ checkFunctions verifiedModule =
     checkResults eqns =
         mapM_ errorEquationRightFunction eqns $> ExitFailure 3
 
+{- | Returns true when both equations match the same term.  See:
+ - https://github.com/kframework/kore/issues/2472#issue-833143685
+-}
+bothMatch ::
+    MonadSimplify m =>
+    Equation VariableName ->
+    Equation VariableName ->
+    m Bool
+bothMatch eq1 eq2 =
+    let pre1 = Equation.requires eq1
+        pre2 = Equation.requires eq2
+        arg1 = fromMaybe Predicate.makeTruePredicate $ Equation.argument eq1
+        arg2 = fromMaybe Predicate.makeTruePredicate $ Equation.argument eq2
+        prio1 = fromMaybe Predicate.makeTruePredicate $ Equation.antiLeft eq1
+        prio2 = fromMaybe Predicate.makeTruePredicate $ Equation.antiLeft eq2
+        check =
+            Predicate.makeAndPredicate pre1 $
+                Predicate.makeAndPredicate pre2 $
+                    Predicate.makeAndPredicate arg1 $
+                        Predicate.makeAndPredicate arg2 $
+                            Predicate.makeAndPredicate prio1 prio2
+        check' = Predicate.mapVariables (pure mkConfigVariable) check
+        sort = termLikeSort $ right eq1
+        patt = Pattern.fromPredicateSorted sort check'
+     in (not . isBottom) <$> Pattern.simplify patt
+
+{- | Checks if any function definition in the module carries two equations that both match
+ - same term.
+-}
+checkBothMatch ::
+    MonadSimplify m =>
+    -- | The main module
+    VerifiedModule StepperAttributes ->
+    m ExitCode
+checkBothMatch verifiedModule =
+    filterM (uncurry bothMatch) equations
+        >>= checkResults
+  where
+    equations = join $ map mkPairs $ Map.elems $ extractEquations verifiedModule
```

Comment:
I just realized something: I think `equations` will contain simplification equations as well! We want to only run these checks (the one you're working on and the one David worked on) for _function definition_ equations.

First, to make sure that this is indeed the case, I'd add a test for each check where we make sure we ignore simplification equations. Remember, simplification equations are those which have the `simplification` attribute. Then, if those tests fail, we need to extract only those equations which _don't_ have the `simplification` attribute, here and for the first check as well. The tests should pass after this.